Documentation/rfc/v3api.md

```diff
@@ -0,0 +1,476 @@
+  // and generates one event in the event history.
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+
+  // Tnx processes all the requests in one transaction.
+  // A tnx request increases the index of the store,
+  // and generates one event in the event history.
+  rpc Tnx(TnxRequest) returns (TnxResponse) {}
+
+  // Watch watches the events happening or happened in etcd. Both input and output
+  // are stream. One watch rpc can watch for multiple ranges and get a stream of
+  // events. The whole events history can be watched unless compacted.
+  rpc Watch(stream WatchRequest) returns (stream WatchResponse) {}
+
+  // Compact compacts the event history in etcd. User should compact the
+  // event history periodically, or it will grow infinitely.
+  rpc Compact(CompactRequest) returns (CompactResponse) {}
+
+  // LeaseCreate creates a lease. A lease has a TTL. The lease will expire if the
+  // server does not receive a keepAlive within TTL from the lease holder.
+  // All keys attached to the lease will be expired and deleted if the lease expires.
+  // The key expiration generates an event in event history.
+  rpc LeaseCreate(LeaseCreateRequest) returns (LeaseCreateResponse) {}
+
+  // LeaseRevoke revokes a lease. All the key attached to the lease will be expired and deleted.
+  rpc LeaseRevoke(LeaseRevokeRequest) returns (LeaseRevokeResponse) {}
+
+  // LeaseAttach attaches keys with a lease.
+  rpc LeaseAttach(LeaseAttachRequest) returns (LeaseAttachResponse) {}
+
+  // LeaseTnx likes Tnx. It does one additional thing that all the keys put into the
+  // store are attached with the given lease. This is rpc is useful when you want to
+  // put a key and attach it to a lease atomically.
+  rpc LeaseTnx(LeaseTnxRequest) returns (LeaseTnxResponse) {}
+
+  // KeepAlive keeps the lease alive.
+  rpc LeaseKeepAlive(stream LeaseKeepAliveRequest) returns (stream LeaseKeepAliveResponse) {}
+}
+
+message RequestHeader {
+  // the cluster id to check with
+  // if the cluster id is not matched, an error is returned.
+  optional uint64 cluster_id = 1;
+}
+
+message ResponseHeader {
+  // an error type message?
+  optional string error = 1;
+  optional uint64 cluster_id = 2;
+  optional uint64 member_id = 3;
+  // index of the store when the request was applied.
+  optional int64 index = 4;
+  // term of raft when the request was applied.
+  optional uint64 raft_term = 5;
+}
+
+message RangeRequest {
+  optional bytes key = 1;
+  // if the end_key is not given, it is a get.
+  // if the end_key is given, it gets the keys in range [key, end_key).
+  optional bytes end_key = 2;
+  // limit the number of keys returned.
+  optional int64 limit = 3;
+  // if not all the keys are returned, a continue id will be 
+  // returned in the range response if consistent is set.
+  // the returned continue id only works with the previous 
+  // etcdserver the client connected to. 
+  optional bool consistent = 4
```

Comment:
----- Original Message -----
> > +  // term of raft when the request was applied.
> > +  optional uint64 raft_term = 5;
> > +}
> > +
> > +message RangeRequest {
> > +  optional bytes key = 1;
> > +  // if the end_key is not given, it is a get.
> > +  // if the end_key is given, it gets the keys in range [key, end_key).
> > +  optional bytes end_key = 2;
> > +  // limit the number of keys returned.
> > +  optional int64 limit = 3;
> > +  // if not all the keys are returned, a continue id will be
> > +  // returned in the range response if consistent is set.
> > +  // the returned continue id only works with the previous
> > +  // etcdserver the client connected to.
> > +  optional bool consistent = 4
> 
> @smarterclayton
> 
> How about giving me an example when you want to use consistent range query
> over multiple requests?
> 
> The only use case I can image is when a server restarts and want to
> reconstruct its in memory state via etcd in a consistent manner then catch
> up with watch (go through the event history).

Right now in Kubernetes a lot of components are aggressively reconstructing their state (via a list) every few minutes.  I think as the 3.0 features land (especially long histories) that interval will decrease.  However, I expect a lot of infrastructure components to want to perform lists on whole ranges (up to a few 10k of records) to perform maintenance and cleanup.  For instance, pruning images from an image repository, where the image metadata is stored in etcd.  Pruning needs to load the entire list of images and the entire list of things that can hold references to those images, compare them, then throw them away.  This operation might run every 15-30 minutes, but would load a lot of keys and a lot of ranges.  Being able to do a snapshot query over a few tens of seconds would make this operation much easier for clients to write.

I think a lot of snapshot read arguments boil down to - it's easier for the programmer.  In the Kube case, there are probably going to be a lot of quasi-admin style operators who want to write queries on the whole cluster, but don't have the sophistication to handle isolation problems with their query (or even know that's a possibility).  Having an easy snapshot range read that the client can perform (and again, this could all be implemented in the client libraries) makes it safer to let cluster admins implement those patterns.  That's the strongest argument I can make right now, and I don't think it *requires* server snapshot read.

> 
> During the normal operations, if a app is operating on a large range, it
> should just watch.
> 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/coreos/etcd/pull/2675/files#r28843797